Library/OcVirtualFsLib/VirtualDir.c

```diff
@@ -0,0 +1,586 @@
+  IN UINTN                    BufferSize,
+  IN VOID                     *Buffer
+  )
+{
+  //
+  // Virtual directories are not writeable, this applies to info.
+  //
+  return EFI_WRITE_PROTECTED;
+}
+
+STATIC
+EFI_STATUS
+EFIAPI
+VirtualDirFlush (
+  IN EFI_FILE_PROTOCOL        *This
+  )
+{
+  //
+  // Virtual directories are not writeable.
+  //
+  return EFI_WRITE_PROTECTED;
+}
+
+STATIC
+EFI_STATUS
+EFIAPI
+VirtualDirOpenEx (
+  IN     EFI_FILE_PROTOCOL    *This,
+  OUT    EFI_FILE_PROTOCOL    **NewHandle,
+  IN     CHAR16               *FileName,
+  IN     UINT64               OpenMode,
+  IN     UINT64               Attributes,
+  IN OUT EFI_FILE_IO_TOKEN    *Token
+  )
+{
+  EFI_STATUS         Status;
+
+  //
+  // Ignore asynchronous interface for now.
+  //
+  // WARN: Unlike Open for OpenEx UEFI 2.7A explicitly dicates EFI_NO_MEDIA for
+  //  "The specified file could not be found on the device." error case.
+  //  We do not care for simplicity.
+  //
+  Status = VirtualDirOpen (
+    This,
+    NewHandle,
+    FileName,
+    OpenMode,
+    Attributes
+    );
+
+  if (!EFI_ERROR (Status) && Token->Event != NULL) {
+    Token->Status = EFI_SUCCESS;
+    gBS->SignalEvent (Token->Event);
+  }
+
+  return Status;
+}
+
+STATIC
+EFI_STATUS
+EFIAPI
+VirtualDirReadEx (
+  IN EFI_FILE_PROTOCOL      *This,
+  IN OUT EFI_FILE_IO_TOKEN  *Token
+  )
+{
+  EFI_STATUS         Status;
+
+  Status = VirtualDirRead (This, Token->Buffer, &Token->BufferSize);
+
+  if (!EFI_ERROR (Status) && Token->Event != NULL) {
+    Token->Status = EFI_SUCCESS;
+    gBS->SignalEvent (Token->Event);
+  }
+
+  return Status;
+}
+
+STATIC
+EFI_STATUS
+EFIAPI
+VirtualDirWriteEx (
+  IN EFI_FILE_PROTOCOL      *This,
+  IN OUT EFI_FILE_IO_TOKEN  *Token
+  )
+{
+  //
+  // Directories are not writeable.
+  //
+  return EFI_UNSUPPORTED;
+}
+
+STATIC
+EFI_STATUS
+EFIAPI
+VirtualDirFlushEx (
+  IN EFI_FILE_PROTOCOL      *This,
+  IN OUT EFI_FILE_IO_TOKEN  *Token
+  )
+{
+  //
+  // Virtual directories are not writeable.
+  //
+  return EFI_WRITE_PROTECTED;
+}
+
+STATIC
+CONST
+EFI_FILE_PROTOCOL
+mVirtualDirProtocolTemplate = {
+  .Revision    = EFI_FILE_PROTOCOL_REVISION2,
+  .Open        = VirtualDirOpen,
+  .Close       = VirtualDirClose,
+  .Delete      = VirtualDirDelete,
+  .Read        = VirtualDirRead,
+  .Write       = VirtualDirWrite,
+  .GetPosition = VirtualDirGetPosition,
+  .SetPosition = VirtualDirSetPosition,
+  .GetInfo     = VirtualDirGetInfo,
+  .SetInfo     = VirtualDirSetInfo,
+  .Flush       = VirtualDirFlush,
+  .OpenEx      = VirtualDirOpenEx,
+  .ReadEx      = VirtualDirReadEx,
+  .WriteEx     = VirtualDirWriteEx,
+  .FlushEx     = VirtualDirFlushEx
+};
+
+EFI_STATUS
+VirtualDirCreateOverlay (
+  IN     CHAR16             *FileName,
+  IN     EFI_TIME           *ModificationTime OPTIONAL,
+  IN     EFI_FILE_PROTOCOL  *UnderlyingFile OPTIONAL,
+  IN OUT EFI_FILE_PROTOCOL  **File
+  )
+{
+  VIRTUAL_DIR_DATA  *Data;
+
+  ASSERT (FileName != NULL);
```

Comment:
Just in case. Unlike in protocol functions, in library code we want `ASSERT` indeed. Please do **not** change `ASSERT` from here onwards.